_posts/2024-09-25-Testing-Scientific-Software.md

```diff
@@ -0,0 +1,374 @@
+        assert is_even(number)
+    else:
+        assert not is_even(number)
+```
+
+### Using the Hypothesis ghostwriter
+
+Hypothesis also includes [a "ghostwriter" CLI](https://hypothesis.readthedocs.io/en/latest/ghostwriter.html) which can infer how to write Hypothesis tests given an object from Python code.
+This can help automate the process of writing your test code or provide inspiration for how to construct your Hypothesis tests.
+_Caveat emptor_: please be sure to review any code generated by the Hypothesis ghostwriter (it may not capture useful tests or edge cases).
+
+Given the example code from the above, we could ask the Hypothesis ghostwriter to construct a test for the `is_even()` function as follows:
+
+```bash
+# run the command from your activated Python environment
+hypothesis write package_name.module.is_even
+
+# or, run through a poetry environment
+poetry run hypothesis write package_name.module.is_even
+```
+
+The output looks similar but not quite the same as the Hypothesis test we shared above.
+Note that the test name implies and itself employs a technique called [fuzzing, or fuzz testing](https://en.wikipedia.org/wiki/Fuzzing), which is used to help determine where software might break.
+
+```python
+# This test code was written by the `hypothesis.extra.ghostwriter` module
+# and is provided under the Creative Commons Zero public domain dedication.
+
+import package_name.module
+from hypothesis import given, strategies as st
+
+
+@given(number=st.integers())
+def test_fuzz_is_even(number: int) -> None:
+    package_name.module.is_even(number=number)
+```
+
+## Benefits of using Hypothesis
+
+Below are just a few of the benefits you'll find with using Hypothesis:
+
+- ___Discovering Edge Cases___: Hypothesis automatically generates diverse input scenarios, including edge cases that might be overlooked in example-based tests.
+- ___Reduced Boilerplate Code___: You can focus on the properties of your functions rather than writing extensive examples for every possible case.
+- ___Increased Confidence___: By validating the behavior of your code against a broader set of inputs, you can be more confident that your scientific software will behave correctly in practice.
+
+## Best practices for scientific software testing
+
+Now that you understand some testing tools, here are some best practices for testing scientific software:
+
+- __Write tests early__: Incorporate testing from the start. 
+This is crucial, especially when your software are prone to evolving.
+
+- __Test small and test often__: Focus on unit tests that cover individual functions and methods.
+Catching small errors early prevents larger problems down the line.
+
+- __Use realistic test data__: When testing your functions, always use test data that reflects the real-world conditions where your software will be applied. Avoid the temptation to use "mock" data which can sometimes be challenging to create and might not accurately test capabilities.
```

Comment:
I somewhat disagree here: while it's good to have a test suite that does a comprehensive, realistic test against a full dataset, it could end up taking a long time or lots of resources, and unfortunately when tests become cumbersome they get disabled and things get missed.

I'd argue that you should have at least two test suites: one that runs quickly and does a pretty good job of testing the software, then another that performs a full, comprehensive test against real data. You'd use the first suite for day-to-day development, then use the second for releases or for more sweeping changes. The "small" suite might use mocks if it's not possible to get it to run quickly without them, or just test a subset of functionality that doesn't require integrating with things you'd mock.

Reply:
Thanks - I felt torn here. Generally I wanted to advise to "avoid until you have to use mocks" to keep the data as real as you can without trying to create or manage synthetics. I'll make some updates here to this effect.